HW4.py

```diff
@@ -0,0 +1,360 @@
+    X = df.data
+    y = df.target
+    response_type = classify_type(y)
+
+    predictor_names = []
+    predictor_types = []
+    plots = []
+    t_values = []
+    p_values = []
+    diff_mean_plots = []
+    diff_mean_tables = []
+
+    # classify predictors as categorical or continuous
+    for idx, column in enumerate(X.T):
+        pred_name = df.feature_names[idx]
+        pred_type = classify_type(column)
+        predictor_names.append(pred_name)
+        predictor_types.append(pred_type)
+
+    # loop through predictors are do appropriate analysis for each type
+    for idx, pred in enumerate(predictor_names):
+        x_type = predictor_types[idx]
+        y_type = response_type
+        x_data = X.T[idx]
+        response = y
+        pred_name = pred
+        if x_type == "cat" and y_type == "cat":
+            plot_link = plot_cat_response_cat_predictor(x_data, pred_name, response)
+            t_value = 0
+            p_value = 0
+        elif x_type == "cat" and y_type == "con":
+            plot_link = plot_con_resp_cat_predictor(x_data, pred_name, response)
+            t_value = 0
+            p_value = 0
+        elif x_type == "con" and y_type == "cat":
+            plot_link = plot_cat_resp_con_predictor(x_data, pred_name, response)
+            t_value, p_value = log_regression(x_data, pred_name, response)
+        elif x_type == "con" and y_type == "con":
+            plot_link = plot_con_response_con_predictor(x_data, pred_name, response)
+            t_value, p_value = lin_regression(x_data, pred_name, response)
+        plot2_link = plot_difference_with_mean(x_data, y, 10, pred_name)
+        table_link = table_difference_with_mean(x_data, y, 10, pred_name)
+        plots.append(plot_link)
+        t_values.append(t_value)
+        p_values.append(p_value)
+        diff_mean_plots.append(plot2_link)
+        diff_mean_tables.append(table_link)
+
+    if response_type == "cat":
+        rank = random_forest_importance_ranking_classify(X, y, len(predictor_names))
+    elif response_type == "con":
+        rank = random_forest_importance_ranking_regressor(X, y, len(predictor_names))
+
+    # put plots, tables, etc in table for final report
+    df = pd.DataFrame(
+        list(
+            zip(
+                predictor_names,
+                predictor_types,
+                plots,
+                t_values,
+                p_values,
+                diff_mean_plots,
+                diff_mean_tables,
+                rank,
+            )
+        ),
+        columns=[
+            "Name",
+            "Type",
+            "Plot",
+            "t-score",
+            "p-value",
+            "diff mean plots",
+            "diff mean tables",
+            "rank",
+        ],
+    )
+    df_sorted = df.sort_values("rank")
+    df_final = df_sorted.style.format(
+        {"Plot": fun, "diff mean plots": fun, "diff mean tables": fun}
+    )
+
+    final_report = df_final.to_html()
```

Comment:
Report looks good, but adding table lines/other styling features may make it look better (unfortunately it's pretty unintuitive to use for pandas Styler objects, there may be better ways I'm missing).  I found code for this from [this](https://stackoverflow.com/questions/64812819/pandas-dataframe-style-lost-table-border) Stack overflow question.